### PR TITLE
[node] Refactor node tests for better ganache/contracts handling support

### PR DESCRIPTION
**Pull request by cf19drofxots**
### Description

This refactors how the test environment gets setup to better support ganache / contract dependencies.

Since jest can't pass state between the `globalSetup` script and individual test suite, the workaround is to deploy the needed contracts in this script and write their addresses to a temp file that gets read by the `testEnvironment` script (which is sandboxed per test suite).

This will better support the upcoming PR for testing Node functionality when proper multisig contracts need to be deployed.

Caveat: As of Jest 24, `transform` can be run in `global{Setup, Teardown}` but not on the `testEnvironment` file, meaning the firebase service factory can't be used in the `testEnvironment` script (as it's TS, and the `testEnvironment` script is JS).

- [x] Deploy preview is functional

### CODE REVIEW

File: packages/node/package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
   "license": "MIT",
   "scripts": {
     "build": "rollup -c",
-    "test": "jest --setupFiles dotenv-extended/config --runInBand --detectOpenHandles --bail",
```
Comment (cf19drofxots) [inline packages/node/package.json]: This is intended until a follow up PR investigates proper closing of handles.

File: packages/node/package.json [hunk 2]
```diff
@@ -52,6 +53,9 @@
         "statements": 90
       }
     },
+    "globalSetup": "<rootDir>/test/global-setup.jest.ts",
```
Comment (snario) [inline packages/node/package.json]: Maybe extract this stuff into `jest.config.js` if its getting more complicated

File: packages/node/src/services.ts [hunk 3]
```diff
@@ -46,6 +46,10 @@ export class FirebaseServiceFactory {
   createStoreService(storeServiceKey: string): IStoreService {
     return new FirebaseStoreService(this.app.database(), storeServiceKey);
   }
+
+  async close() {
```
Comment (snario) [inline packages/node/src/services.ts]: Maybe `closeConnection` or `disconnect` or `disconnectAndDeleteHistory` (does it delete history?)

File: packages/node/test/global-setup.jest.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import dotenvExtended from "dotenv-extended";
+import { ContractFactory, Wallet } from "ethers";
+import { Web3Provider } from "ethers/providers";
+import fs from "fs";
+import ganache from "ganache-core";
+import mkdirp from "mkdirp";
+import os from "os";
+import path from "path";
+
+dotenvExtended.load();
```
Comment (snario) [inline packages/node/test/global-setup.jest.ts]: Isn't the `--setupFiles dotenv-extended/config` thing supposed do this?

File: packages/node/test/global-setup.jest.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        balance: "120000000000000000",
+        secretKey: process.env.PRIVATE_KEY_B
+      },
+      {
+        balance: "120000000000000000",
+        secretKey: process.env.PRIVATE_KEY_C
+      }
+    ]
+  });
+  // @ts-ignore
```
Comment (snario) [inline packages/node/test/global-setup.jest.ts]: Reasoning behind this?

File: packages/node/test/global-setup.jest.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  const proxyFactoryContract = await new ContractFactory(
+    ProxyFactory.abi,
+    ProxyFactory.bytecode,
+    wallet
+  ).deploy();
+  const tttContract = await new ContractFactory(
+    TicTacToeApp.interface,
+    TicTacToeApp.bytecode,
+    wallet
+  ).deploy();
```
Comment (snario) [inline packages/node/test/global-setup.jest.ts]: Perhaps this counterfactual-specific stuff should be in a separate file than the ganache stuff? Just to keep this file simple and easy to grok

File: packages/node/test/node-test-environment.jest.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    super(config);
+  }
+
+  async setup() {
+    await super.setup();
+    const addresses = readFileSync(path.join(DIR, "addresses"), "utf8");
+    if (!addresses) {
+      throw new Error("Contract addresses not found");
+    }
+    this.global.contractAddresses = JSON.parse(addresses);
```
Comment (snario) [inline packages/node/test/node-test-environment.jest.js]: Maybe stay consistent with the `networkContext` variable name

File: packages/node/test/global-setup.jest.ts (hunk 4 above)
Comment (cf19drofxots) [reply packages/node/test/global-setup.jest.ts]: It apparently doesn't work on setup files.

I assumed it did on my first try but the values ended up being `undefined`.

File: packages/node/test/global-setup.jest.ts (hunk 5 above)
Comment (cf19drofxots) [reply packages/node/test/global-setup.jest.ts]: TS complains that a property being set on the `Global` object doesn't exist (`Property 'ganacheServer' does not exist on type 'Global'`) even though expected behavior for Jest is to be able to set properties like this.

So ignoring the false TS warning.

**Pull request merged**